### PR TITLE
feat: add ops that miss in the tosa standard and add linalg conversion for them

### PR DESCRIPTION
**Pull request by maxbartel**
I added a new dialect "namespace" for this extension. A follow up story might be to rename the quantization ops to this scheme.

Also added new Traits for a bit of a taxonomy when working with them. The `ToseExtension Trait` can be converted to an interface if the TOSA decomposition is needed.

### CODE REVIEW

File: include/xten/Dialect/XTenNN/IR/XTenNNOps.td [hunk 1]
```diff
@@ -181,4 +181,26 @@ def XTenNN_LoadExternalConstOp: XTenNN_Op<"load_external_const", [
...
 }
 
+//===----------------------------------------------------------------------===//
+// Ops that are missing from the TOSA standard
+//===----------------------------------------------------------------------===//
+
+def TosaExtension : NativeOpTrait<"TosaExtension">;
+def ElementwiseUnary : NativeOpTrait<"ElementwiseUnary">;
+
+def XtenNN_UnifiedSignOp: XTenNN_Op<"unified.sign", [Pure, TosaExtension, ElementwiseUnary, SameOperandsAndResultElementType]> {
```
Comment (mgehre-amd) [inline include/xten/Dialect/XTenNN/IR/XTenNNOps.td]: What does `unified` mean here?

Comment (maxbartel) [reply include/xten/Dialect/XTenNN/IR/XTenNNOps.td]: Because it is unified from onnx and torch. Naming is hard....

Comment (mgehre-amd) [reply include/xten/Dialect/XTenNN/IR/XTenNNOps.td]: I feel like `xten_nn.sign` is good enough

**Comment by maxbartel** [comment]
@mgehre-amd I will add a commit to mention that the base for the linalg conversion comes from stablehlo

**Comment by mgehre-amd** [comment]
Do torch and onnx share the same semantics with sign on NaN/+inf/-inf? Does our implementation preserve those semantics?

File: include/xten/Dialect/XTenNN/IR/XTenNNOps.td [hunk 2]
```diff
@@ -181,4 +181,26 @@ def XTenNN_LoadExternalConstOp: XTenNN_Op<"load_external_const", [
...
+  let summary = "Calculate the sign of the given input tensor element-wise";
+  let description = [{
+    Calculate the sign of the given input tensor element-wise. If input > 0, output 1. if input < 0, output -1. if input == 0, output 0.
+  }];
```
Comment (mgehre-amd) [inline include/xten/Dialect/XTenNN/IR/XTenNNOps.td:196]: Please mention NaN/+inf/-inf here

File: include/xten/Dialect/XTenNN/IR/XTenNNOps.td (hunk 1 above)
Comment (maxbartel) [reply include/xten/Dialect/XTenNN/IR/XTenNNOps.td]: But it is not the same class of op like the subgraph for example. I feel like some kind of separation needs to happen here. I even thought about a new dialect

File: test/Conversion/XTenNNToLinalg/Sign.mlir [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// CHECK-SAME:                        %[[VAL_0:.*]]: tensor<1x10xi4>) -> tensor<1x10xi4> {
+// CHECK:           %[[VAL_1:.*]] = tensor.empty() : tensor<1x10xi4>
+// CHECK:           %[[VAL_2:.*]] = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%[[VAL_0]] : tensor<1x10xi4>) outs(%[[VAL_1]] : tensor<1x10xi4>) {
+// CHECK:           ^bb0(%[[VAL_3:.*]]: i4, %[[VAL_4:.*]]: i4):
+// CHECK:             %[[VAL_5:.*]] = arith.constant 0 : i4
+// CHECK:             %[[VAL_6:.*]] = arith.constant 3 : i4
+// CHECK:             %[[VAL_7:.*]] = arith.constant 1 : i4
+// CHECK:             %[[VAL_8:.*]] = arith.cmpi eq, %[[VAL_3]], %[[VAL_5]] : i4
+// CHECK:             %[[VAL_9:.*]] = arith.shrsi %[[VAL_3]], %[[VAL_6]] : i4
+// CHECK:             %[[VAL_10:.*]] = arith.ori %[[VAL_9]], %[[VAL_7]] : i4
```
Comment (mgehre-amd) [inline test/Conversion/XTenNNToLinalg/Sign.mlir:37]: Why is this an or? shouldn't this be applying a mask via arith.and?

**Comment by maxbartel** [comment]
> Do torch and onnx share the same semantics with sign on NaN/+inf/-inf? Does our implementation preserve those semantics?

The onnx implementation is a direct copy of the torch one as far as I understand. I can try to look that up, the onnx reference models are implement in python though, not sure how well defined they will be

File: test/Conversion/XTenNNToLinalg/Sign.mlir (hunk 3 above)
Comment (maxbartel) [reply test/Conversion/XTenNNToLinalg/Sign.mlir:37]: Hmm think of a -3 which is 1101 in binary. The signed shift will output 1111 and the same after the or. This is -1 in two's complement, so the sign. Why do you want an and here?

File: lib/Conversion/XTenNNToLinalg.cpp [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    Value zero =
+        b->create<arith::ConstantOp>(loc, b->getZeroAttr(operand.getType()));
+    Value bitwidthMinusOne = getConstantOrSplat(
+        b, loc, operand.getType(),
+        b->getIntegerAttr(integerType, integerType.getWidth() - 1));
+    Value one = getConstantOrSplat(b, loc, operand.getType(),
+                                   b->getIntegerAttr(integerType, 1));
+    Value cmp = b->create<::mlir::arith::CmpIOp>(loc, arith::CmpIPredicate::eq,
+                                                 operand, zero);
+    Value ashr =
```
Comment (mgehre-amd) [inline lib/Conversion/XTenNNToLinalg.cpp:94]: Now I get what you are doing. Could you please add a comment to explain it to others?

**Review by mgehre-amd — APPROVED** [review]

**Pull request merged**